### PR TITLE
Fix bug with small sorts and shared scans, introduced by 8.3 merge.

### PR DESCRIPTION
**Pull request by hlinnaka**
The old GPDB code had a "fall through" in switch-case block, that was not
present in PostgreSQL 8.2. PostgreSQL 8.3 added a new case-block to the
statement, so that the GPDB-specific "fall through" fell through to the
wrong case-block. This lead to a crash, if a sort was small enough to be
performed in memory, was shared by multiple slices. Fix by moving the
GPDB-specific check outside the switch-case statement.

### CODE REVIEW

File: src/backend/utils/sort/tuplesort.c [hunk 1]
```diff
@@ -1370,34 +1370,33 @@ tuplesort_performsort(Tuplesortstate *state)
 		elog(LOG, "performsort starting: %s",
 			 pg_rusage_show(&state->ru_start));
 
+	/*
+	 * If we the output needs to be shared by other readers, we need to
```
Comment (asimrp) [inline src/backend/utils/sort/tuplesort.c]: Typo: s/we//

**Comment by asimrp** [comment]
Does a test already exist in ICG that exercises this part of the code?

**Comment by hlinnaka** [comment]
On 12/13/2016 06:01 PM, Asim RP wrote:
> Does a test already exist in ICG that exercises this part of the code?

I got a crash because of this while running installcheck-good. But I 
haven't seen it before and the bug's been there since the 8.3 merge. I'm 
not sure what suddenly made it appear on my system, and I'm not sure how 
to reproduce it, I'm afraid.

- Heikki

**Comment by hlinnaka** [comment]
I just got the segfault again, while hacking on something else. The failed query was this:
```
SELECT avg(vn) OVER (PARTITION BY cn) FROM ow_sale;
```
from qp_olap_window test. And I think I now  know why the concourse pipeline hasn't shown this failure: I have gp_enable_mk_sort=off in the config file. That was leftover from some earlier testing I did with the cluster. Don't remember what exactly that was for, but it's clearly useful to test non-default config options every now and then :-).

**Comment by asimrp** [comment]
The test inserts small enough data into `ow_sale` table, meaning that the sort will always happen in memory.  Shouldn't it always crash?  I'm not able to reason the indeterminate nature of the crash.

**Comment by hlinnaka** [comment]
On 14 December 2016 20:02:28 EET, Asim RP <notifications@github.com> wrote:
>The test inserts small enough data into `ow_sale` table, meaning that
>the sort will always happen in memory.  Shouldn't it always crash?  I'm
>not able to reason the indeterminate nature of the crash.

"set gp_enable_mk_sort=off" does the trick. With that, it always crashes. It's not the default, so it's not exercised by the CI pipeline.

- Heikki

**Comment by hlinnaka** [comment]
Here's a full standalone repro, extracted from qp_window_olap.sql:
```
SET gp_enable_sequential_window_plans = off;
SET gp_enable_mk_sort = off;

create table ow_sale
(
        cn int not null,
        vn int not null,
        pn int not null,
        dt date not null,
        qty int not null,
        prc float not null,
        
        primary key (cn, vn, pn)
        
) distributed by (cn,vn,pn);

SELECT avg(vn) OVER (PARTITION BY cn) FROM ow_sale;
```

**Comment by kuien** [comment]
Affirmative, this sql can repro the case: FailedAssertion() on my Macbook too.

**Comment by asimrp** [comment]
👍 thank you for the exact repro.

**Pull request merged**